index.d.ts

```diff
@@ -129,6 +131,41 @@ export namespace Cmd {
   }
   export type PromiseResult<T> = T extends Promise<infer U> ? U : T;
 
+  export function run<Func extends (...args: any[]) => Promise<any> | any>(
+    f: Func,
+    options?: {
+      args?: ArgOrSymbol<Parameters<Func>>;
+      forceSync?: boolean;
+      testInvariants?: boolean;
+    },
+  ): RunCmd;
+
+  export function run<
+    Func extends (...args: any[]) => Promise<any> | any,
+    SuccessAction extends Action = never,
```

Comment:
shouldn't there be no default here? Otherwise it's just the same as the first overload

Reply:
Oh yeah, that's a good catch. It was a remnant from the original, non-overload approach I think. Ditto for the failure action default - the point of the overload is that if you're using it that you have one.

Reply:
While we're on the topic though... I actually don't know how / if TS could tell the difference between these two overloads.

```
  export function run<
    Func extends (...args: any[]) => Promise<any> | any,
    SuccessAction extends Action,
    >(
    f: Func,
    options?: {
      args?: ArgOrSymbol<Parameters<Func>>;
      successActionCreator: (value: PromiseResult<ReturnType<Func>>) => SuccessAction;
      forceSync?: boolean;
      testInvariants?: boolean;
    },
  ): RunCmd<SuccessAction, never>;

  export function run<
    Func extends (...args: any[]) => Promise<any> | any,
    FailAction extends Action,
    >(
    f: Func,
    options?: {
      args?: ArgOrSymbol<Parameters<Func>>;
      failActionCreator: (error: any) => FailAction;
      forceSync?: boolean;
      testInvariants?: boolean;
    },
  ): RunCmd<never, FailAction>;
```

They both take the same number of generics and the generics are the same type. I'm betting that this won't work.. We probably need to delete both of those and just have one signature that takes no actions and one that takes two actions.

```
  export function run<Func extends (...args: any[]) => Promise<any> | any>(
    f: Func,
    options?: {
      args?: ArgOrSymbol<Parameters<Func>>;
      forceSync?: boolean;
      testInvariants?: boolean;
    },
  ): RunCmd;

  export function run<
    Func extends (...args: any[]) => Promise<any> | any,
    SuccessAction extends Action,
    FailAction extends Action,
    >(
    f: Func,
    options?: {
      args?: ArgOrSymbol<Parameters<Func>>;
      failActionCreator: (error: any) => FailAction;
      successActionCreator: (value: PromiseResult<ReturnType<Func>>) => SuccessAction;
      forceSync?: boolean;
      testInvariants?: boolean;
    },
  ): RunCmd<SuccessAction, FailAction>;
```

Thoughts?

Reply:
I pushed up what I'm thinking in commit https://github.com/redux-loop/redux-loop/pull/244/commits/c54b476487d23f7532fe9851c5eb0a47da1b304c.

Reply:
it can because one takes a failActionCreator and the other takes a success.